src/containers/CustomFunctionModal.tsx

```diff
@@ -0,0 +1,212 @@
+import parser, { getNames } from '../parser'
+import type { CustomFunction } from '../reducers/customFunctions'
+
+interface CustomFunctionModalProps {
+  customFunction: CustomFunction
+  deactivateModal: () => void
+  createFunction: (name: string, func: string) => void
+  setFunction: (id: string, name: string, func: string) => void
+  deleteFunction: (id: string) => void
+}
+
+interface CustomFunctionModalState {
+  name: string
+  func: string
+  nameError: string
+  funcError: string
+}
+
+class CustomFunctionModal extends Component<CustomFunctionModalProps, CustomFunctionModalState> {
+  nameRef: React.RefObject<HTMLInputElement>
+
+  funcRef: React.RefObject<HTMLTextAreaElement>
+
+  constructor(props: CustomFunctionModalProps) {
+    super(props)
+    const { customFunction: cf } = this.props
+
+    this.nameRef = createRef()
+    this.funcRef = createRef()
+    this.state = {
+      name: cf ? cf.name : '',
+      func: cf ? cf.func : '',
+      nameError: '',
+      funcError: '',
+    }
+  }
+
+  componentDidMount() {
+    if (this.nameRef.current) {
+      this.nameRef.current.focus()
+    }
+  }
+
+  validateName = () => {
+    const { name } = this.state
+    if (!name) {
+      this.setState({ nameError: t`Please name your function` })
+      return false
+    }
+    return true
+  }
+
+  validateFunc = () => {
+    const { func } = this.state
+    let functionVariables: string[] = []
+    const setGenericError = () => this.setState({ funcError: t`There was an error with your function.` })
+
+    try {
+      functionVariables = getNames(func)
+    } catch (err) {
+      console.error(err)
+      setGenericError()
+      return false
+    }
+
+    if (functionVariables.length === 0) {
+      this.setState({ funcError: t`You need to include at least one variable in your function.` })
+      return false
+    }
+
+    const variableNames = variables.map(v => v.name)
+    const invalidVariables: string[] = []
+    functionVariables.forEach(v => {
+      if (![...variableNames, 'math_e', ...invalidVariables].includes(v)) {
+        invalidVariables.push(v)
+      }
+    })
+    if (invalidVariables.length === 1) {
+      this.setState({ funcError: `"${invalidVariables[0]}" ${t`is not a known variable`}` })
+      return false
+    }
+    if (invalidVariables.length > 1) {
+      this.setState({ funcError: `"${invalidVariables.join(', ')}" ${t`are not known variables`}` })
+      return false
+    }
+
+    // sets each variable to `2` and parses function to catch parsing errors
+    try {
+      const context: any = {}
+      functionVariables.forEach(variable => {
+        context[variable] = 2
+      })
+      parser(func, context)
+    } catch (err) {
+      console.error(err)
+      setGenericError()
+      return false
+    }
+
+    return true
+  }
+
+  onSave = () => {
+    const { customFunction, deactivateModal, createFunction, setFunction } = this.props
+    const { validateName, validateFunc } = this
```

Comment:
I think it's fine to just use `this.validateName()` and `this.validateFunc()` when calling the functions. That's more common than assigning methods to local variables first.